src/snowflake/snowpark/functions.py

```diff
@@ -3209,6 +3209,105 @@ def array_intersection(array1: ColumnOrName, array2: ColumnOrName) -> Column:
+        >>> df1 = session.create_dataframe([(-2, 2)], ["a", "b"])
+        >>> df1.select(array_generate_range("a", "b").alias("result")).show()
+        ------------
+        |"RESULT"  |
+        ------------
+        |[         |
+        |  -2,     |
+        |  -1,     |
+        |   0,     |
+        |   1      |
+        |]         |
+        ------------
+        <BLANKLINE>
+        >>> df2 = session.create_dataframe([(4, -4, -2)], ["a", "b", "c"])
+        >>> df1.select(array_generate_range("a", "b", "c").alias("result")).show()
+        ------------
+        |"RESULT"  |
+        ------------
+        |[         |
+        |   4,     |
+        |   2,     |
+        |   0,     |
+        |  -2      |
+        |]         |
+        ------------
+        <BLANKLINE>
+    """
+    start_col = _to_col_if_str(start, "array_generate_range")
+    stop_col = _to_col_if_str(stop, "array_generate_range")
+    if step is None:
+        return builtin("array_generate_range")(start_col, stop_col)
+    step_col = _to_col_if_str(step, "array_generate_range")
+    return builtin("array_generate_range")(start_col, stop_col, step_col)
+
+
+def sequence(
+    start: ColumnOrName, stop: ColumnOrName, step: Optional[ColumnOrName] = None
+) -> Column:
+    """Generate a sequence of integers from `start` to `stop`, incrementing by `step`.
+    If `step` is not set, incrementing by 1 if start is less than or equal to stop, otherwise -1.
+
+    Args:
+        start: the column that contains the integer to start with (inclusive).
+        stop: the column that contains the integer to stop (inclusive).
+        step: the column that contains the integer to increment.
+
+    Example::
+        >>> from snowflake.snowpark import Row
+        >>> df1 = session.create_dataframe([(-2, 2)], ["a", "b"])
+        >>> df1.select(sequence("a", "b").alias("result")).show()
+        ------------
+        |"RESULT"  |
+        ------------
+        |[         |
+        |  -2,     |
+        |  -1,     |
+        |   0,     |
+        |   1,     |
+        |   2      |
+        |]         |
+        ------------
+        <BLANKLINE>
+        >>> df2 = session.create_dataframe([(4, -4, -2)], ["a", "b", "c"])
+        >>> df1.select(sequence("a", "b", "c").alias("result")).show()
+        ------------
+        |"RESULT"  |
+        ------------
+        |[         |
+        |   4,     |
+        |   2,     |
+        |   0,     |
+        |  -2,     |
+        |  -4      |
+        |]         |
+        ------------
+        <BLANKLINE>
+    """
+    start_col = _to_col_if_str(start, "sequence")
+    stop_col = _to_col_if_str(stop, "sequence")
+    if step is None:
+        step = iff(builtin("sign")(stop_col - start_col) > 0, 1, -1)
+        return builtin("array_generate_range")(start_col, stop_col + step, step)
+    step_col = _to_col_if_str(step, "sequence")
+    return builtin("array_generate_range")(start_col, stop_col + step_col, step_col)
```

Comment:
I think it should be `stop_col + 1`? `step_col` can be larger than one, so we might include number larger than `stop_col`. For example, if we have `sequence(0, 5, 4)`, my understanding is that we should only return `[0, 4]`, but the current implementation is going to return `[0, 4, 8]`.

Reply:
very good catch thanks!